content/blog/2025-03-11-Alive-Peephole-Optimizations

```diff
@@ -0,0 +1,62 @@
+
+Six years after the release of this paper, in 2021, Alive was updated to the tool Alive 2, which expands Alive to include loop handling (by partially unrolling loops) and goes beyond simple local optimizations to handle function calls as well. It also uses translation validation which takes two IR files and checks whether the translation from one to the other is correct, so that the optimization can be tried on a test case before it is implemented. Alive 2 also covers floats and aggregate types, which were missing from the implementation of the original Alive.
+
+This new version of Alive, Alive2, has essentially replaced Alive’s usefulness in the LLVM development sphere, but it is still in use today and the contributions of Alive were essential in facilitating its creation.
+
+## Criticisms/Concerns:
+The authors do a lot to justify the usefulness of their tool; the anecdote about finding bugs and then eventually verifying correctness in a real developer’s actual LLVM patch is particularly effective. Reading the article, one can’t help but feel bad for that poor embarrassed developer for having bugs caught in the first two editions of their PR, but it is a convincing argument to use the tool; certainly from a software development perspective, you’d rather use Alive in-house and catch those bugs yourself, rather than send your patch out into the world and have someone else embarrass you by catching the bugs in your code!
+
+We also think it makes a lot of sense to design Alive with LLVM in mind, to find an already thriving developer community and create a tool specifically for them, rather than create something radically new and different that people are then resistant to actually use in real life. Tools with a concrete audience in mind are usually better loved and more impactful overall than even the most groundbreaking tool that people resist using. One can argue that since Alive can be rather difficult to understand for people without an LLVM background, there is some loss in impact with broader audiences for the language, but this is not really a shortcoming.
+
+However, there were a lot of concerns raised in the online discussion about both the scope/usefulness and correctness of the DSL. For example, with the scope of Alive only covering Peephole optimizations, which are small, simple cases that can be easily hand-checked, the sentiment kept popping up: how useful is the tool, really? There was a lot of interest in trying to expand a verification tool like Alive to more complex optimizations. However, through our discussion in class, we came to the conclusion that this kind of expansion was difficult to realistically achieve, and there was some dissent that this level of optimization was useless; the evidence that the paper presents about catching bugs in real-life LLVM programs using their tool suggests its usefulness, even if it only focuses on these small local peephole optimizations.
```

Comment:
> Peephole optimizations, which are small, simple cases that can be easily hand-checked

I would argue that the paper provides plenty of examples where peephole optimizations _cannot_ feasibly be hand-checked, or at least cases where humans got them wrong despite attempting to hand-check them.

Reply:
Similarly, please steer clear of "useless": I don't think it's worth representing the position that verifying peephole optimizations is truly useless. I think what you mean here is that there may be other transformations that matter even more (despite the fact that InstCombine is disproportionately buggy, pound for pound).